lib/plugins/aws/package/compile/events/apiGateway/index.js

```diff
@@ -20,15 +21,256 @@ const getMethodAuthorization = require('./lib/method/authorization');
+  properties: {
+    cors: corsSchema,
+    method: { type: 'string', regexp: methodPattern.toString() },
+    path: { type: 'string', regexp: /^(?:\*|\/\S*)$/.toString() },
+    operationId: { type: 'string' },
+    async: { type: 'boolean' },
+    private: { type: 'boolean' },
+    request: requestSchema,
+    response: responseSchema,
+    authorizer: authorizerSchema,
+  },
+  required: ['path', 'method'],
+  additionalProperties: false,
+};
+
+const lambdaProxyIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: caseInsensitive('lambda_proxy'),
+  },
+  required: ['integration'],
+});
+
+const mockIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: caseInsensitive('mock'),
+  },
+  required: ['integration'],
+});
+
+const awsIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      oneOf: ['aws', 'aws_proxy'].map(caseInsensitive),
+    },
+  },
+  required: ['integration'],
+});
+
+const httpIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      oneOf: ['http', 'http_proxy'].map(caseInsensitive),
+    },
+    request: {
+      required: ['uri'],
+    },
+  },
+  required: ['integration'],
+});
+
+const httpVpcLinkIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    connectionType: { enum: ['vpc-link', 'VPC_LINK'] },
+    connectionId: { type: 'string' },
+    integration: {
+      oneOf: ['http', 'http_proxy'].map(caseInsensitive),
+    },
+    request: {
+      required: ['uri'],
+    },
+  },
+  required: ['integration', 'connectionType', 'connectionId'],
+});
+
+const lambdaIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: caseInsensitive('lambda'),
+    authorizer: {
+      properties: {
+        claims: { type: 'array', items: { type: 'string' } },
+      },
+    },
+  },
+  required: ['integration'],
+});
+
+const cognitoIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: caseInsensitive('cognito_user_pools'),
+    authorizer: {
+      required: ['name'],
+    },
+  },
+  required: ['integration'],
+});
+
 class AwsCompileApigEvents {
   constructor(serverless, options) {
     this.serverless = serverless;
     this.options = options;
     this.provider = this.serverless.getProvider('aws');
 
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'http', {
-      // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8018
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      oneOf: [
+        { type: 'string', regexp: methodPathPattern.toString() },
+        baseIntegrationSchema,
+        awsIntegrationSchema,
+        cognitoIntegrationSchema,
+        httpIntegrationSchema,
+        httpVpcLinkIntegrationSchema,
+        lambdaIntegrationSchema,
+        lambdaProxyIntegrationSchema,
+        mockIntegrationSchema,
```

Comment:
One problem with this kind of configuration is that eventual errors will not be informative.

afaik if I declare `method: FOO` in otherwise perfectly valid configuration, the result error will be: _"Unsupported configuration format"_, with no hint that there's a value in `method`.

It goes from fact, that with such schema it's challenging to resolve which exactly `oneOf` option was chosen by user. AJV will expose errors resolved for each option, showing all of them will produce unbearable noise (we intentionally convert it then, to one _"Unsupported configuration format"_)

We already had similar cases before, and we decided to configure properties for all options in one level, and leave (or introduce) inline validation that's meaningful for user as e.g. follows:

```javascript

if (integrationType !== X && authorizer.claims) {
  throw new ServelressError("authorizer.claims is not supported for given integration type");
}
```

I would also stick here to that.

I believe we should not try to validate any complex interchangeable rules with JSON schema, as it then it becomes a maintenance burden we can't show a meaningful error message to user.

Reply:
I understand.
I will refactor this PR to have a single schema, and leave the validations needed for the edge cases.